net: Enhance support for v4 mapping to v6

Cristib05 opened the pull request:
This commit aims to enhance v4 mapping to v6 functionality.

- Fixed an issue when `sendmsg()` was in use and `CONFIG_NET_IPV6` and `CONFIG_NET_IPV4_MAPPING_TO_IPV6` were in use, as `dst_addr` argument of `context_sendto()` was passed NULL, but also used to determine `sa_family`.
- Updated `set_pkt_hoplimit` function to take into account v4 to v6 mapping.
- Updated `sock_get_pkt_src_addr` to return mapped address.

Enahncement to #63579
Related to #94156

Cristib05 commented:
Hi @jukkar. Can you please let me know your thoughts on this? Thanks!

jukkar commented on subsys/net/ip/net_context.c:
```diff
@@ -2345,12 +2350,21 @@ static void set_pkt_hoplimit(struct net_pkt *pkt, const struct msghdr *msg_hdr)
-				net_pkt_set_ipv4_ttl(pkt, *(uint8_t *)CMSG_DATA(cmsg));
-				break;
+		} else if (net_pkt_family(pkt) == AF_INET) {
+			if (!addr6 || (addr6 && !net_ipv6_addr_is_v4_mapped(&addr6->sin6_addr))) {
```
Please change `!addr6` to `addr6 == NULL` as we try to comply to some Misra rule that says the short variant should only used if the variable is boolean.
Also change `addr6 &&` to `addr != NULL &&`

jukkar commented on subsys/net/ip/net_context.c:
```diff
@@ -2395,8 +2409,16 @@ static int context_sendto(struct net_context *context,
 	if (IS_ENABLED(CONFIG_NET_IPV4_MAPPING_TO_IPV6) &&
 	    IS_ENABLED(CONFIG_NET_IPV6) &&
 	    net_context_get_family(context) == AF_INET6 &&
+	    dst_addr &&
```
`dst_addr != NULL &&`

jukkar commented on subsys/net/ip/net_context.c:
```diff
@@ -2395,8 +2409,16 @@ static int context_sendto(struct net_context *context,
 	    dst_addr->sa_family == AF_INET) {
 		family = AF_INET;
+	} else if (IS_ENABLED(CONFIG_NET_IPV4_MAPPING_TO_IPV6) &&
+		   IS_ENABLED(CONFIG_NET_IPV6) && msghdr) {
```
`msghdr != NULL`

jukkar commented on subsys/net/ip/net_context.c:
```diff
@@ -2335,6 +2335,11 @@ static void set_pkt_txtime(struct net_pkt *pkt, const struct msghdr *msghdr)
 	struct cmsghdr *cmsg;
+	const struct sockaddr_in6 *addr6 = NULL;
+
+	if (IS_ENABLED(CONFIG_NET_IPV4_MAPPING_TO_IPV6) && IS_ENABLED(CONFIG_NET_IPV6)) {
```
Don't we need to check that the `pkt` addr family is `AF_INET6` here? This function could be called for IPv4 too.

jukkar replied:
Ah, my bad, was reading the diff badly, sorry for the noise.

jukkar reviewed (commented):
Overall, LGTM

jukkar commented:
You have three changes in the PR, could you create three commits too as the changes are somewhat separate

Cristib05 commented:
> You have three changes in the PR, could you create three commits too as the changes are somewhat separate

ok, I will split them

Cristib05 commented:
> BTW, what do you mean by `This commit aims to enhance v4 mapping to v6 for UDP communication.` I did not notice any UDP specific changes in the PR.

You are right, I've did this while developing border router application which makes use of UDP. This is why I think I've mentioned UDP here. Will update

Cristib05 committed:
net: context: Add support to set hoplimit when v4 to v6 mapping is used

If v4 to v6 mapping is used, cmsg_level and cmsg_type should
be compared with IPv6 values even if pkt family if AF_INET.

Signed-off-by: Cristian Bulacu <cristian.bulacu@nxp.com>

Cristib05 committed:
net: context: Fix issues in context_sendto() function

Fixed issue when `sendmsg()` was in use and `CONFIG_NET_IPV6`
and `CONFIG_NET_IPV4_MAPPING_TO_IPV6` were in use, as `dst_addr`
argument of `context_sendto()` was passed NULL,
but also used to determine `sa_family`.
Use msghdr, if available, to determine if v4 to v6 mapping is
present.
Also set `dst_addr` to unmapped value  when v4 to v6 mapping is
used.

Signed-off-by: Cristian Bulacu <cristian.bulacu@nxp.com>

Cristib05 committed:
net: sockets: Return mapped address from `sock_get_pkt_src_addr()` function

When an IPv4 packet is received on a dual-stack socket, return
the mapped v4 to v6 address as source address.

Signed-off-by: Cristian Bulacu <cristian.bulacu@nxp.com>

Cristib05 commented:
For the moment, I am not able to reproduce reported twister errors on my setup. I could reproduce the previous ones.

Cristib05 committed:
tests: net: socket: tls: Increase test stack size

Increased test stack size to avoid twister failures.

Signed-off-by: Cristian Bulacu <cristian.bulacu@nxp.com>
